src/mempool/block_receiver.rs

```diff
@@ -0,0 +1,201 @@
+use prost::Message;
+use std::time::Duration;
+use tokio::select;
+use tokio::sync::{broadcast, mpsc, oneshot};
+use tokio::time::Instant;
+use tracing::{info, warn};
+
+use crate::consensus::consensus::SystemMessage;
+use crate::mempool::mempool::{MempoolRequest, MempoolSource};
+use crate::proto::{hub_event, Block, HubEvent};
+use crate::storage::store::{mempool_poller::MempoolMessage, stores::Stores};
+use thiserror::Error;
+
+#[derive(Error, Debug)]
+pub enum BlockReceiverError {
+    #[error("timed out waiting for confirmation")]
+    ConfirmationTimedOut,
+}
+// Maintain one per shard, shards move independently
+pub struct BlockReceiver {
+    pub shard_id: u32,
+    pub block_rx: broadcast::Receiver<Block>,
+    pub mempool_tx: mpsc::Sender<MempoolRequest>,
+    pub system_tx: mpsc::Sender<SystemMessage>,
+    pub event_rx: broadcast::Receiver<HubEvent>,
+    pub stores: Stores,
+}
+
+impl BlockReceiver {
+    fn validate_block_events(&self, block: &Block) -> bool {
+        if block.events.len() == 0 {
+            return true;
+        }
+
+        let mut events_hasher = blake3::Hasher::new();
+        for event in &block.events {
+            if event.data.is_none() {
+                return false;
+            }
+
+            if event.hash
+                != blake3::hash(&event.data.as_ref().unwrap().encode_to_vec())
+                    .as_bytes()
+                    .to_vec()
+            {
+                return false;
+            }
+
+            events_hasher.update(&event.hash);
+        }
+
+        if block.header.as_ref().unwrap().events_hash
+            != events_hasher.finalize().as_bytes().to_vec()
+        {
+            return false;
+        }
+
+        // TODO(aditi): Validate signatures
```

Comment:
It was not designed for this, so might be annoying to refactor but we should see if we can re-use the code from the block proposer. This logic is important enough that we shouldn't risk introducing divergences by duplicating the code

Reply:
Good point, reused.